// @langtrase/trace-attributes/dist/constants/common.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TIKTOKEN_MODEL_MAPPING = exports.TracedFunctionsByVendor = exports.Event = exports.Vendors = exports.FrameworkSpanAttributeNames = exports.DatabaseSpanAttributeNames = exports.LLMSpanAttributeNames = void 0;
const llm_span_attributes_json_1 = require("../schemas/llm_span_attributes.json");
const database_span_attributes_json_1 = require("../schemas/database_span_attributes.json");
const framework_span_attributes_json_1 = require("../schemas/framework_span_attributes.json");
const anthropic_1 = require("./anthropic");
const pg_1 = require("./pg");
const chroma_1 = require("./chroma");
const cohere_1 = require("./cohere");
const groq_1 = require("./groq");
const llamaindex_1 = require("./llamaindex");
const openai_1 = require("./openai");
const pinecone_1 = require("./pinecone");
const qdrant_1 = require("./qdrant");
const weaviate_1 = require("./weaviate");
const ollama_1 = require("./ollama");
const ai_1 = require("./ai");
exports.LLMSpanAttributeNames = Object.keys(llm_span_attributes_json_1.properties);
exports.DatabaseSpanAttributeNames = Object.keys(database_span_attributes_json_1.properties);
exports.FrameworkSpanAttributeNames = Object.keys(framework_span_attributes_json_1.properties);
exports.Vendors = {
    OPENAI: 'openai',
    COHERE: 'cohere',
    ANTHROPIC: 'anthropic',
    GROQ: 'groq',
    PINECONE: 'pinecone',
    LLAMAINDEX: 'llamaindex',
    CHROMADB: 'chromadb',
    QDRANT: 'qdrant',
    WEAVIATE: 'weaviate',
    PG: 'pg',
    VERCEL: 'ai',
    OLLAMA: 'ollama'
};
var Event;
(function (Event) {
    Event["STREAM_START"] = "stream.start";
    Event["STREAM_OUTPUT"] = "stream.output";
    Event["STREAM_END"] = "stream.end";
    Event["RESPONSE"] = "response";
    Event["GEN_AI_COMPLETION"] = "gen_ai.content.completion";
    Event["GEN_AI_COMPLETION_CHUNK"] = "gen_ai.completion.chunk";
    Event["GEN_AI_PROMPT"] = "gen_ai.content.prompt";
})(Event || (exports.Event = Event = {}));
exports.TracedFunctionsByVendor = {
    anthropic: anthropic_1.AnthropicFunctionNames,
    pg: pg_1.PgFunctionNames,
    chromadb: chroma_1.ChromadbFunctionNames,
    cohere: cohere_1.CohereFunctionNames,
    groq: groq_1.GroqFunctionNames,
    llamaindex: llamaindex_1.LlamaIndexFunctionNames,
    openai: openai_1.OpenAIFunctionNames,
    pinecone: pinecone_1.PineConeFunctionNames,
    qdrant: qdrant_1.QdrantFunctionNames,
    weaviate: weaviate_1.WeaviateFunctionNames,
    ai: ai_1.VercelAIFunctionNames,
    ollama: ollama_1.OllamaFunctionNames
};
exports.TIKTOKEN_MODEL_MAPPING = {
    'gpt-4': 'cl100k_base',
    'gpt-4-32k': 'cl100k_base',
    'gpt-4-0125-preview': 'cl100k_base',
    'gpt-4-1106-preview': 'cl100k_base',
    'gpt-4-1106-vision-preview': 'cl100k_base',
    'gpt-4o': 'o200k_base',
    'gpt-4o-2024-05-13': 'o200k_base'
};
