// @langtrase/typescript-sdk/dist/instrumentation/weaviate/instrumentation.js
"use strict";
/*
 * Copyright (c) 2024 Scale3 Labs
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.weaviateInstrumentation = void 0;
const instrumentation_1 = require("@opentelemetry/instrumentation");
// eslint-disable-next-line no-restricted-imports
const package_json_1 = require("../../../package.json");
const api_1 = require("@opentelemetry/api");
const patch_1 = require("../weaviate/patch");
class WeaviateInstrumentation extends instrumentation_1.InstrumentationBase {
    constructor() {
        super(package_json_1.name, package_json_1.version);
    }
    manualPatch(weaviate) {
        api_1.diag.debug('Manually instrumenting weaviate');
        this._patch(weaviate);
    }
    init() {
        const module = new instrumentation_1.InstrumentationNodeModuleDefinition('weaviate-ts-client', ['>=2.2.0'], (moduleExports, moduleVersion) => {
            api_1.diag.debug(`Patching weaviate SDK version ${moduleVersion}`);
            this._patch(moduleExports, moduleVersion);
            return moduleExports;
        }, (moduleExports, moduleVersion) => {
            api_1.diag.debug(`Unpatching weaviate SDK version ${moduleVersion}`);
            if (moduleExports !== undefined) {
                this._unpatch(moduleExports);
            }
        });
        return [module];
    }
    _patch(weaviate, moduleVersion) {
        this._wrap(weaviate.default, 'client', (original) => {
            return (params) => {
                const clientInstance = original.apply(this, [params]);
                patch_1.patchBuilderFunctions.call(this, {
                    clientInstance,
                    clientArgs: params,
                    tracer: this.tracer,
                    moduleVersion: moduleVersion,
                    sdkName: package_json_1.name,
                    sdkVersion: package_json_1.version
                });
                return clientInstance;
            };
        });
    }
    _unpatch(weaviate) {
        this._unwrap(weaviate.default, 'client');
    }
}
exports.weaviateInstrumentation = new WeaviateInstrumentation();
