// @langtrase/trace-attributes/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.queryTypeToFunctionToProps = exports.TIKTOKEN_MODEL_MAPPING = exports.APIS = exports.Vendors = exports.TracedFunctionsByVendor = exports.Event = exports.FrameworkSpanAttributeNames = exports.DatabaseSpanAttributeNames = exports.LLMSpanAttributeNames = void 0;
const common_1 = require("./constants/common");
Object.defineProperty(exports, "LLMSpanAttributeNames", { enumerable: true, get: function () { return common_1.LLMSpanAttributeNames; } });
Object.defineProperty(exports, "DatabaseSpanAttributeNames", { enumerable: true, get: function () { return common_1.DatabaseSpanAttributeNames; } });
Object.defineProperty(exports, "FrameworkSpanAttributeNames", { enumerable: true, get: function () { return common_1.FrameworkSpanAttributeNames; } });
Object.defineProperty(exports, "Event", { enumerable: true, get: function () { return common_1.Event; } });
Object.defineProperty(exports, "TracedFunctionsByVendor", { enumerable: true, get: function () { return common_1.TracedFunctionsByVendor; } });
Object.defineProperty(exports, "Vendors", { enumerable: true, get: function () { return common_1.Vendors; } });
const anthropic_1 = require("./constants/anthropic");
const pg_1 = require("./constants/pg");
const chroma_1 = require("./constants/chroma");
const cohere_1 = require("./constants/cohere");
const groq_1 = require("./constants/groq");
const llamaindex_1 = require("./constants/llamaindex");
const openai_1 = require("./constants/openai");
const pinecone_1 = require("./constants/pinecone");
const qdrant_1 = require("./constants/qdrant");
const ai_1 = require("./constants/ai");
const ollama_1 = require("./constants/ollama");
const weaviate_1 = require("./constants/weaviate");
Object.defineProperty(exports, "queryTypeToFunctionToProps", { enumerable: true, get: function () { return weaviate_1.queryTypeToFunctionToProps; } });
const common_2 = require("./constants/common");
Object.defineProperty(exports, "TIKTOKEN_MODEL_MAPPING", { enumerable: true, get: function () { return common_2.TIKTOKEN_MODEL_MAPPING; } });
const APIS = {
    anthropic: anthropic_1.APIS,
    pg: pg_1.APIS,
    chromadb: chroma_1.APIS,
    cohere: cohere_1.APIS,
    groq: groq_1.APIS,
    llamaindex: llamaindex_1.APIS,
    openai: openai_1.APIS,
    pinecone: pinecone_1.APIS,
    qdrant: qdrant_1.APIS,
    ai: ai_1.APIS,
    ollama: ollama_1.APIS
};
exports.APIS = APIS;
